pkg/operator2/datasync/validation.go

```diff
@@ -0,0 +1,136 @@
+package datasync
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	corelistersv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/util/keyutil"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+var validators = map[string]func(data []byte) []error{
+	corev1.TLSCertKey:       validateClientCert,
+	corev1.TLSPrivateKeyKey: validatePrivateKey,
+
+	corev1.ServiceAccountRootCAKey: validateCACerts,
+	configv1.ClientSecretKey:       noValidation,
+	configv1.HTPasswdDataKey:       noValidation,
+	configv1.BindPasswordKey:       noValidation,
+}
+
+func noValidation(_ []byte) []error { return []error{} }
+
+func validateSecret(secretsLister corelistersv1.SecretLister, src sourceData) []error {
+	s, err := secretsLister.Secrets("openshift-config").Get(src.Src)
+	if err != nil {
+		return []error{err}
+	}
+
+	data, exists := s.Data[src.Key]
+	if !exists {
+		return []error{fmt.Errorf("missing required key: %q", src.Key)}
+	}
+
+	return validators[src.Key](data)
```

Comment:
Would it make sense to check that the key was found and return an error if not?

Reply:
I don't mind the panic in case the programmer forgets to add a validator